scipy/optimize/shgo_m/sobol_seq.py

```diff
@@ -0,0 +1,371 @@
+          |   16 |      10000 |   1
+          |   17 |      10001 |   2
+          | 1023 | 1111111111 |   1
+          | 1024 | 0000000000 |   1
+          | 1025 | 0000000001 |   1
+
+        Parameters:
+          Input, integer N, the integer to be measured.
+          N should be nonnegative.
+
+          Output, integer BIT, the position of the low 1 bit.
+        """
+        bit = 1
+        i = np.floor(n)
+        while i != 2 * (i // 2):
+            bit += 1
+            i //= 2
+        return bit
+
+    def i4_sobol(self, dim_num, seed):
+        """
+        i4_sobol generates a new quasirandom Sobol vector with each call.
+
+        Discussion:
+          The routine adapts the ideas of Antonov and Saleev.
+
+        Reference:
+          Antonov, Saleev,
+          USSR Computational Mathematics and Mathematical Physics,
+          Volume 19, 1980, pages 252 - 256.
+
+          Paul Bratley, Bennett Fox,
+          Algorithm 659:
+          Implementing Sobol's Quasirandom Sequence Generator,
+          ACM Transactions on Mathematical Software,
+          Volume 14, Number 1, pages 88-100, 1988.
+
+          Bennett Fox,
+          Algorithm 647:
+          Implementation and Relative Efficiency of Quasirandom
+          Sequence Generators,
+          ACM Transactions on Mathematical Software,
+          Volume 12, Number 4, pages 362-376, 1986.
+
+          Ilya Sobol,
+          USSR Computational Mathematics and Mathematical Physics,
+          Volume 16, pages 236-242, 1977.
+
+          Ilya Sobol, Levitan,
+          The Production of Points Uniformly Distributed in a Multidimensional
+          Cube (in Russian),
+          Preprint IPM Akad. Nauk SSSR,
+          Number 40, Moscow 1976.
+
+        Parameters:
+          Input, integer DIM_NUM, the number of spatial dimensions.
+          DIM_NUM must satisfy 1 <= DIM_NUM <= 40.
+
+          Input/output, integer SEED, the "seed" for the sequence.
+          This is essentially the index in the sequence of the quasirandom
+          value to be generated.  On output, SEED has been set to the
+          appropriate next value, usually simply SEED+1.
+          If SEED is less than 0 on input, it is treated as though it were 0.
+          An input value of 0 requests the first (0-th) element of the sequence.
+
+          Output, real QUASI(DIM_NUM), the next quasirandom vector.
+        """
+
+        # if 'self.initialized' not in list(globals().keys()):
+        if self.initialized is None:
+            self.initialized = 0
+            self.dim_num_save = -1
+
+        if not self.initialized or dim_num != self.dim_num_save:
+            self.initialized = 1
+            self.dim_max = 40
+            self.dim_num_save = -1
+            self.log_max = 30
+            self.seed_save = -1
+
+            #  Initialize (part of) V.
+            self.v = np.zeros((self.dim_max, self.log_max))
+            self.v[0:40, 0] = np.transpose([
```

Comment:
In NumPy slicing notation you don't need to declare the end like matlab and transpose is not needed since transpose of 1D array in NumPy is itself.

```
self.v = np.zeros((self.dim_max, self.log_max), dtype=int)
self.v[:, 0] = [1]*self.dim_max
self.v[2:, 1] = ([1, 3]*3 + ([3, 1]*4 + [1, 3]*4)*2)
self.v[3:, 2] = [7, 5, 1, 3, 3, 7, 5,
                 5, 7, 7, 1, 3, 3, 7, 5, 1, 1,
                 5, 3, 3, 1, 7, 5, 1, 3, 3, 7,
                 5, 1, 1, 5, 7, 7, 5, 1, 3, 3]
```
and so on.

Also 40 is now hard-coded which should be `dim_max` I think